File: excercise1/main.go
package main

import (
	"fmt"
	"sync"
	"./test"
)

var i int = 0

func main() {
	var wg sync.WaitGroup
	wg.Add(2)
	go thread_1(&wg)
	go thread_2(&wg)
	wg.Wait()
	fmt.Println("i is now:", i)
}


func thread_1(wg *sync.WaitGroup) {
	defer wg.Done()
	for j := 0; j < 1000000; j++ {
		i++
	}
}

func thread_2(wg *sync.WaitGroup) {
	defer wg.Done()
	for j := 0; j < 1000000; j++ {
		i--
	}
}

File: excercise1/main.c
#include <stdio.h>
#include <stdlib.h>
#include <pthread.h>

int i = 0;
void *increment( void *ptr ) {
    for (int j = 0; j < 1000000; j++)
        i++;
}
void *decrement( void *ptr ) {
    for (int j = 0; j < 1000000; j++)
        i--;
}


int main()
{
     pthread_t thread1, thread2;
     int  iret1, iret2;

    /* Create independent threads each of which will execute function */

     iret1 = pthread_create( &thread1, NULL, increment, NULL);
     iret2 = pthread_create( &thread2, NULL, decrement, NULL);

     /* Wait till threads are complete before main continues. Unless we  */
     /* wait we run the risk of executing an exit which will terminate   */
     /* the process and all threads before the threads have completed.   */

     pthread_join( thread1, NULL);
     pthread_join( thread2, NULL); 

     printf("i is now: %d\n",i);
     exit(0);
}
